File: cpp/hot100/48.路径总和 III.cpp
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        int ans=0;
        unordered_map<long long, int> count{{0,1}};
        auto dfs=[&](this auto&& dfs, TreeNode *node, long long s){
            if(!node) return ;
            s += node->val;
            ans += count[s-targetSum];
            count[s]++;
            dfs(node->left, s);
            dfs(node->right, s);
            count[s]--;
        };
        dfs(root, 0);
        return ans;
    }
};

File: python/hot100/48.路径总和 III.py
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        self.ans = 0
        prefix_sum = defaultdict(int)
        prefix_sum[0] = 1
        def dfs(node, pre_sum):
            if not node:
                return
            
            pre_sum += node.val
            self.ans += prefix_sum[pre_sum-targetSum]
            prefix_sum[pre_sum] += 1
            dfs(node.left, pre_sum)
            dfs(node.right,pre_sum)
            prefix_sum[pre_sum]-=1
        dfs(root, 0)
        return self.ans